### ISSUE TITLE
smallfile not working on windows

### ISSUE THREAD

**Issue by ksandha** [distributed-system-analysis/smallfile#23]
I tried the latest master branch on windows client. It returned me below error

```
adding time for Windows synchronization
Traceback (most recent call last):
  File "C:\Users\administrator.GCHILD\Desktop\smallfile-master\smallfile-master\
smallfile_cli.py", line 288, in <module>
    run_workload()
  File "C:\Users\administrator.GCHILD\Desktop\smallfile-master\smallfile-master\
smallfile_cli.py", line 279, in run_workload
    return multi_thread_workload.run_multi_thread_workload(params)
  File "C:\Users\administrator.GCHILD\Desktop\smallfile-master\smallfile-master\
multi_thread_workload.py", line 121, in run_multi_thread_workload
    sync_files.write_sync_file(sg, 'hi there')
  File "C:\Users\administrator.GCHILD\Desktop\smallfile-master\smallfile-master\
sync_files.py", line 16, in write_sync_file
    os.rename(fpath+notyet, fpath)
WindowsError: [Error 32] The process cannot access the file because it is being
used by another process
```

But branch "smallfile-dir-share-wi-clnts" is working perfectly on single client.

**Comment by bengland2** [comment]
@ksandha thanks for informing me of this, just saw this. I haven't tested on Windows in a long time so this is not surprising, but I'll take a look at it.

**Comment by bengland2** [comment]
@ksandha sorry but I don't have a bunch of Windows clients lying around or easily creatable.   Give me the exact command you used, all the output,  and the inputs to that command, and I'll try to figure out what happened.   Are you using a SMB share or an NFS share?  I'm guessing SMB.

**Comment by bengland2** [comment]
by the way you have to use --launch-by-daemon to run smallfile on Windows, did you do that?

**Comment by bengland2** [comment]
For smallfile to support Windows, we need to create windows VMs and run regression tests in there.   Ideally this would be done automatically, I don't know how to do this yet.

**Comment by sandrobonazzola** [comment]
Tried running test on windows but it fails trying to load libc (which is kind of expected to fail IMHO):

```python
  File "D:\a\smallfile\smallfile\yaml_parser.py", line 13, in <module>
    import smallfile
  File "D:\a\smallfile\smallfile\smallfile.py", line 70, in <module>
    import drop_buffer_cache
  File "D:\a\smallfile\smallfile\drop_buffer_cache.py", line 40, in <module>
    _posix_fadvise = load_libc_function('posix_fadvise64')
  File "D:\a\smallfile\smallfile\drop_buffer_cache.py", line [30](https://github.com/sandrobonazzola/smallfile/actions/runs/3313449095/jobs/5471441642#step:6:31), in load_libc_function
    libc = ctypes.CDLL(ctypes.util.find_library('c'))
```

Testing with GitHub action: https://github.com/sandrobonazzola/smallfile/commit/2aeab3226021ba13b315f961593b28995ab629f8

**Comment by portante** [comment]
Perhaps make "drop cache" behavior a no-op on platforms that don't support it?

**Comment by sandrobonazzola** [comment]
unit testing on windows are passing with 71% coverage according to https://github.com/distributed-system-analysis/smallfile/actions/runs/3433776816

---

# Coverage report: 71%
coverage.py v6.5.0, created at 2022-11-10 03:39 +0000 

Module | statements | missing | excluded | coverage  
---|---|---|---|---  
drop_buffer_cache.py | 35 | 15 | 0 | 57%  
fallocate.py | 37 | 17 | 0 | 54%  
invoke_process.py | 98 | 18 | 0 | 82%  
parser_data_types.py | 49 | 27 | 0 | 45%  
smallfile.py | 1450 | 339 | 0 | 77%  
smf_test_params.py | 98 | 69 | 0 | 30%  
sync_files.py | 47 | 21 | 0 | 55%  
yaml_parser.py | 146 | 57 | 0 | 61%  
Total | 1960 | 563 | 0 | 71%  
  
coverage.py v6.5.0, created at 2022-11-10 03:39 +0000 

---

I think this issue can be closed. @portante what do you think?

**Comment by portante** [comment]
Fixed by PR #39.
